contracts/liquidity_hub/bonding-manager/src/contract.rs

```diff
@@ -0,0 +1,295 @@
+use crate::state::{BONDING_ASSETS_LIMIT, CONFIG, EPOCHS};
+use crate::{commands, queries};
+
+// version info for migration info
+const CONTRACT_NAME: &str = "white_whale-bonding_manager";
+const CONTRACT_VERSION: &str = env!("CARGO_PKG_VERSION");
+
+#[entry_point]
+pub fn instantiate(
+    deps: DepsMut,
+    _env: Env,
+    info: MessageInfo,
+    msg: InstantiateMsg,
+) -> Result<Response, ContractError> {
+    if msg.bonding_assets.len() > BONDING_ASSETS_LIMIT {
+        return Err(ContractError::InvalidBondingAssetsLimit(
+            BONDING_ASSETS_LIMIT,
+            msg.bonding_assets.len(),
+        ));
+    }
+
+    validate_growth_rate(msg.growth_rate)?;
+
+    set_contract_version(deps.storage, CONTRACT_NAME, CONTRACT_VERSION)?;
+
+    let config = Config {
+        owner: deps.api.addr_validate(info.sender.as_str())?,
+        pool_manager_addr: Addr::unchecked(""),
+        distribution_denom: msg.distribution_denom,
+        unbonding_period: msg.unbonding_period,
+        growth_rate: msg.growth_rate,
+        bonding_assets: msg.bonding_assets.clone(),
+        grace_period: msg.grace_period,
+    };
+
+    CONFIG.save(deps.storage, &config)?;
+    // Creates a new bucket for the rewards flowing from this time on, i.e. to be distributed in the next epoch. Also, forwards the expiring epoch (only 21 epochs are live at a given moment)
+    // Add a new rewards bucket for the new epoch
+    EPOCHS.save(
+        deps.storage,
+        &0u64.to_be_bytes(),
+        &Epoch {
+            id: 0u64.into(),
+            start_time: _env.block.time,
+            ..Epoch::default()
+        },
+    )?;
+    Ok(Response::default().add_attributes(vec![
+        ("action", "instantiate".to_string()),
+        ("owner", config.owner.to_string()),
+        ("unbonding_period", config.unbonding_period.to_string()),
+        ("growth_rate", config.growth_rate.to_string()),
+        ("bonding_assets", msg.bonding_assets.join(", ")),
+        ("grace_period", config.grace_period.to_string()),
+    ]))
+}
+
+#[entry_point]
+pub fn execute(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: ExecuteMsg,
+) -> Result<Response, ContractError> {
+    match msg {
+        ExecuteMsg::Bond {} => {
+            let config = CONFIG.load(deps.storage)?;
+            // Ensure that the user has sent some funds
+            ensure!(!info.funds.is_empty(), PaymentError::NoFunds {});
+            let asset_to_bond = {
+                // Filter the funds to include only those with accepted denominations
+                let valid_funds: Vec<&Coin> = info
+                    .funds
+                    .iter()
+                    .filter(|coin| config.bonding_assets.contains(&coin.denom))
+                    .collect();
+
+                // Check if there are no valid funds after filtering
+                if valid_funds.is_empty() {
+                    Err(PaymentError::MissingDenom("test".to_string()))
```

Comment:
What's this MissingDenom "test"? I think we should return `PaymentError::NoFunds` in that case.